### DISCUSSION TITLE
SQLite reflection misses UNIQUE constraint when there are tabs in the table DDL

### DISCUSSION THREAD

**Discussion by volcan01010** [sqlalchemy/sqlalchemy#11745] [Usage Questions] [answered]
### Summary

I am reflecting the table structure from a SQLite database, but only some of the unique constraints are being picked up. On further investigation, it turns out that the missing ones are where there is a tab (`\t`) in the column definition line after the column name. I think that this is a bug and have proposed a fix, but I am posting here first in case I'm wrong.

### Recreating the problem

The following script creates a SQLite database from a script. Two columns are given UNIQUE constraints, but the definition for the "tab" column has a tab after the column name. The tables are reflected and the constraints are printed. The UNIQUE constraint on the "tab" column was not detected.

```python
# unique_reflection.py
from pathlib import Path
import sqlite3

from sqlalchemy import (
    MetaData,
    create_engine,
)

ddl_script = """
    BEGIN TRANSACTION;
    CREATE TABLE IF NOT EXISTS "test" (
	    "id"	INTEGER NOT NULL,
	    "tab"	TEXT NOT NULL UNIQUE,
	    "space" TEXT NOT NULL UNIQUE,
	    PRIMARY KEY("id" AUTOINCREMENT)
    );
    COMMIT;
"""
print('DDL:')
print(repr(ddl_script))

# Create a new SQLite database with single table
Path('test.db').unlink(missing_ok=True)
with sqlite3.connect('test.db') as conn:
    cursor = conn.cursor()
    cursor.executescript(ddl_script)

# Reflect the table into MetaData class
engine = create_engine(url="sqlite:///test.db")
metadata = MetaData()
metadata.reflect(bind=engine)

# Print the table constraints
print('\nDetected constraints:')
table = metadata.tables["test"]
for constraint in table.constraints:
    print(constraint)
```

The output of the script is:

```
$ python unique_reflection.py
DDL:
'\n    BEGIN TRANSACTION;\n    CREATE TABLE IF NOT EXISTS "test" (\n\t    "id"\tINTEGER NOT NULL,\n\t    "tab"\tTEXT NOT NULL UNIQUE,\n\t    "space" TEXT NOT NULL UNIQUE,\n\t    PRIMARY KEY("id" AUTOINCREMENT)\n    );\n    COMMIT;\n'

Detected constraints:
UniqueConstraint(Column('space', TEXT(), table=<test>, nullable=False))
PrimaryKeyConstraint(Column('id', INTEGER(), table=<test>, primary_key=True, nullable=False))
```

The DDL script where I found the issue was created via SQLite using the `.dump` command, so it is valid code.  SQLite stores the command used to create the table originally in the form that it was provided.  So it is valid to have tabs in the table definitions.

### Proposed fix

This problem is similar to an issue reported in 2022 ([Handle sqlite column names escaped with [ ]](https://github.com/sqlalchemy/sqlalchemy/issues/7736)).
That was fixed by making a change to the regular expression below, which is used to parse the SQLite table definition.

https://github.com/sqlalchemy/sqlalchemy/blob/442657794cc5bcf975b426d7f1104fc25c3b4039/lib/sqlalchemy/dialects/sqlite/base.py#L2590-L2593

A further change the regular expression to the following fixes my use case.

```python
INLINE_UNIQUE_PATTERN = (
    r'(?:(".+?")|(?:[\[`])?([a-z0-9_]+)(?:[\]`])?)[\t ]'
    r"+[a-z0-9_ ]+? +UNIQUE"
)
```

This has replaced ` +` after the part the column name group with `[\t ]+`, so that it matches all tabs and spaces and not just spaces.

#### Notes on fix

+ I chose this, rather than `\s+` because other whitespace characters such as newlines are not valid here.
+ You could do the same with the ` +` before `UNIQUE`, but I have only ever seen tabs used between the column name and the type definition and not between type definition terms.
+ As before, this regex works wherever the UNIQUE command is e.g. with `TEXT UNIQUE NOT NULL`

I would like to make a pull request with the fix, but this change doesn't really fit with the rest of the test (https://github.com/sqlalchemy/sqlalchemy/blob/442657794cc5bcf975b426d7f1104fc25c3b4039/test/dialect/test_sqlite.py#L2460), which is mainly looking at different table name styles.

Having said that, I _think_ that adding the following to the list of test parameters would demonstrate that the fix has worked.

```python
("name_followed_by_tab\t", "name_followed_by_tab"),
```

**Comment by zzzeek** [comment]
sure we just had a PR for almost the same issue with CHECK constraints in #11677 so you can make a PR that is modeled off of that one, which you can see at #11701 .

**Reply by volcan01010** [reply]
OK, will do.

**Comment by CaselIT** [comment]
This should be an issue

**Reply by CaselIT** [reply, marked as answer]
Created #11746
